File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int k, req;
vector<long long> arr;

long long getCount(long long len);

void binary_search() {
    long long lt = 1, rt = *max_element(arr.begin(), arr.end());
    long long res = 0;
    while (lt <= rt) {
        long long m = (lt + rt) / 2;
        if (getCount(m) >= req) {
            res = m; // 최댓값을 갱신
            lt = m + 1;
        } else {
            rt = m - 1;
        }
    }
    cout << res;
}

long long getCount(long long len) {
    long long cnt = 0;
    for (int i = 0; i < arr.size(); i++) {
        cnt += arr[i] / len;
    }
    return cnt;
}

int main() {
    cin >> k >> req;
    arr.resize(k);
    for (int i = 0; i < k; i++) {
        cin >> arr[i];
    }

    binary_search();
    return 0;
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;
import java.util.*;

public class Main {
    static ArrayList<Integer> arr = new ArrayList<>();
    static long n, k, max, mid, res = 0;

    static boolean binary_search(long mid){
        int cnt = 0;
        for (int x : arr){
            cnt += (int) (x / mid);
        }
        return cnt >= n;
    }

    static void solution(){
        long lt = 1;
        long rt = max;
        while (lt <= rt){
            mid = (lt + rt) / 2;
            if(binary_search(mid)) {
                res = Math.max(res,mid);
                lt = mid+1;
            }
            else rt = mid - 1;
        }
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        k = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        for (int i = 0; i < k; i++){
            arr.add(Integer.parseInt(br.readLine()));
            max = Math.max(arr.get(i),max);
        }

        solution();
        System.out.println(res);

    }
}